### PR TITLE
Integrating cfile

### PR DESCRIPTION
**Pull request by itsmontoya**

### CODE REVIEW

File: mrT.go [hunk 1]
```diff
@@ -202,29 +224,10 @@ func (m *MrT) isInCurrent(txnID string) (ok bool) {
...
-	if af, err = os.Open(path.Join(m.dir, "archive", m.name+".tdb")); err != nil {
-		return
-	}
-	defer af.Close()
-
-	as := seeker.New(af)
+	ar := m.af.Reader()
+	defer ar.Close()
+	as := seeker.New(ar)
 	defer as.SetFile(nil)
```
Comment (OneOfOne) [inline mrT.go]: what's the point of this call?

File: mrT.go [hunk 2]
```diff
@@ -415,79 +483,86 @@ func (m *MrT) ForEachTxn(txnID string, archive bool, fn ForEachTxnFn) (err error
...
-		return errors.ErrIsClosed
-	}
+		aw := m.af.Writer()
+		defer aw.Close()
 
-	if lastTxn, err = peekLastTxn(m.s); err != nil {
-		return
-	}
+		// Ensure archive file is at the end
+		if _, err = aw.Seek(0, os.SEEK_END); err != nil {
```
Comment (OneOfOne) [inline mrT.go]: can we use the new io.SeekEnd, etc constants? the os ones are deprecated.

File: mrT.go [hunk 3]
```diff
@@ -677,129 +642,18 @@ func (m *MrT) Export(txnID string, w io.Writer) (err error) {
...
-// peekFirstTxn will return the first transaction id within the current file
-func peekFirstTxn(s *seeker.Seeker) (txnID string, err error) {
-	if err = s.SeekToStart(); err != nil {
-		return
-	}
-
-	return nextTxn(s)
-}
+	// Acquire writers to ensure locks have completed
+	m.f.Writer().Close()
```
Comment (OneOfOne) [inline mrT.go]: calling `m.f.Close()` waits for all the locks.

File: utilities.go [hunk 4]
```diff
@@ -114,3 +118,153 @@ func getLineType(buf *bytes.Buffer) (lineType byte, err error) {
...
+
+func newLBuf() *lbuf {
+	var l lbuf
+	l.buf = bytes.NewBuffer(nil)
+	return &l
+}
+
+type lbuf struct {
+	mux atoms.Mux
+	buf *bytes.Buffer
```
Comment (OneOfOne) [inline utilities.go]: doesn't need to be a pointer.

File: utilities.go [hunk 5]
```diff
@@ -114,3 +118,153 @@ func getLineType(buf *bytes.Buffer) (lineType byte, err error) {
 func endOnMatch(buf *bytes.Buffer) (err error) {
 	return seeker.ErrEndEarly
 }
+
+func newLBuf() *lbuf {
```
Comment (OneOfOne) [inline utilities.go]: just `return &lbuf{}` and don't make bytes.Buffer a pointer ;)

File: mrT.go [hunk 6]
```diff
@@ -55,23 +55,39 @@ func New(dir, name string, mws ...middleware.Middleware) (mp *MrT, err error) {
...
 	// Create new write buffer
-	mrT.buf = bytes.NewBuffer(nil)
+	mrT.lbuf = newLBuf()
+
 	// Create new seeker
-	mrT.s = seeker.New(mrT.f)
+	//	mrT.s = seeker.New(mrT.f)
 	// Set Mr.T's middleware
 	mrT.setMWs(mws)
+	// Set last transaction
```
Comment (shahzilabid) [inline mrT.go]: This should be it's own helper func IMO.. mrT.SetLastTransaction or something

**Review by OneOfOne — COMMENTED** [review]
Great work brother, just few notes.

about the os.SEEK* consts, i just commented in one place but they are all over.

File: mrT.go [hunk 7]
```diff
@@ -415,79 +483,86 @@ func (m *MrT) ForEachTxn(txnID string, archive bool, fn ForEachTxnFn) (err error
...
-		return
-	}
+		if _, err = f.Seek(0, os.SEEK_SET); err != nil {
+			return
+		}
 
-	if _, err = io.Copy(af, m.f); err != nil {
-		return
-	}
+		// TODO: CLEAN UP THIS LOCKCEPTION
```
Comment (shahzilabid) [inline mrT.go]: Lol

File: mrT.go (hunk 1 above)
Comment (itsmontoya) [reply mrT.go]: It's just to clear the file from the seeker for easier garbage collection

File: mrT.go (hunk 2 above)
Comment (itsmontoya) [reply mrT.go]: Oh sure!

File: mrT.go (hunk 1 above)
Comment (OneOfOne) [reply mrT.go]: This is 100% not needed and the overhead of a defer is probably much higher than the uber-tiny-micro-tiny optimization.

File: mrT_test.go [hunk 8]
```diff
@@ -26,23 +27,28 @@ func TestMrT(t *testing.T) {
...
-		txn.Put([]byte("name"), []byte("world"))
+		if err = txn.Put([]byte("greeting"), []byte("hello")); err != nil {
+			return
+		}
+
+		err = txn.Put([]byte("name"), []byte("world"))
 		return
 	}); err != nil {
-		return
+		t.Fatal(err)
```
Comment (OneOfOne) [inline mrT_test.go:37]: t.Fatal will prevent defer from running. please either use t.Panic or just something like:
```
import . github.com/PathDNA/testutils
```
...
```
if FailIf(t, err) { return }
```

Comment (itsmontoya) [reply mrT_test.go:37]: Not true,

```go
func TestDerp(t *testing.T) {
	defer fmt.Println("Hai!")
	t.Fatal("Uhh")
}

/*
panda@archPanda~/go/src/github.com/itsmontoya/mrT: (integrate_cfile) $ go test -run=^TestDerp$
Hai!
--- FAIL: TestDerp (0.00s)
        mrT_test.go:222: Uhh
FAIL
exit status 1
FAIL    github.com/itsmontoya/mrT       0.002s
*/
```

Comment (OneOfOne) [reply mrT_test.go:37]: hmm when did that change. good catch

**Review by OneOfOne — APPROVED** [review]

File: mrT_test.go (hunk 8 above)
Comment (itsmontoya) [reply mrT_test.go:37]: I appreciate you looking out for issues like that though, good shit :+1:

**Review by shahzilabid — APPROVED** [review]
Awesome! Great job!

**Pull request merged**